api/apps/geoprocessing/src/import/pieces-importers/planning-area-custom-grid.piece-importer.ts

```diff
@@ -0,0 +1,142 @@
+      kind === ResourceKind.Project
+    );
+  }
+
+  async run(input: ImportJobInput): Promise<ImportJobOutput> {
+    const { uris, importResourceId, piece } = input;
+    if (uris.length !== 1) {
+      const errorMessage = `uris array has an unexpected amount of elements: ${uris.length}`;
+      this.logger.error(errorMessage);
+      throw new Error(errorMessage);
+    }
+    const [planningAreaCustomGridLocation] = uris;
+
+    const readableOrError = await this.fileRepository.get(
+      planningAreaCustomGridLocation.uri,
+    );
+    if (isLeft(readableOrError)) {
+      const errorMessage = `File with piece data for ${piece}/${importResourceId} is not available at ${planningAreaCustomGridLocation.uri}`;
+      this.logger.error(errorMessage);
+      throw new Error(errorMessage);
+    }
+
+    await this.processGridFile(
+      readableOrError.right,
+      planningAreaCustomGridLocation,
+      importResourceId,
+    );
+
+    return {
+      importId: input.importId,
+      componentId: input.componentId,
+      importResourceId,
+      componentResourceId: input.componentResourceId,
+      piece: input.piece,
+    };
+  }
+
+  private parseGridFileLine(pu: string) {
+    const regex = /^(\d+),(\[(\d+,)*\d+\])$/gi;
+    const result = regex.exec(pu);
+
+    if (result) {
+      const [_, puId, geom] = result;
+      return { puId: parseInt(puId), geom: JSON.parse(geom) as number[] };
+    }
+    throw new Error('unknown line format');
+  }
+
+  private async processGridFile(
+    zipStream: Readable,
+    planningAreaCustomGridLocation: ComponentLocationSnapshot,
+    projectId: string,
+  ) {
+    return new Promise<void>((resolve, reject) => {
+      let lastChunkIncompletedData = '';
+
+      zipStream
+        .pipe(ParseOne(new RegExp(planningAreaCustomGridLocation.relativePath)))
+        .pipe(
+          new Transform({
+            writableObjectMode: true,
+            transform: async (chunk, encoding, callback) => {
+              const data = lastChunkIncompletedData + chunk.toString();
+              const lastNewLineIndex = data.lastIndexOf('\n');
+              if (lastNewLineIndex === -1) {
+                lastChunkIncompletedData = data;
+                callback();
+                return;
+              }
+              const processableData = data.substring(0, lastNewLineIndex);
+              lastChunkIncompletedData = data.substring(lastNewLineIndex + 1);
+              try {
+                const geomPUs = processableData
+                  .split('\n')
+                  .map(this.parseGridFileLine);
+
+                const values = geomPUs
+                  .map((pu, index) => `($1,$2,ST_GeomFromEWKB($${index + 3}))`)
+                  .join(',');
+
+                const buffers = geomPUs.map((pu) => Buffer.from(pu.geom));
+                await this.geoprocessingEntityManager.query(
+                  `
+                  INSERT INTO planning_units_geom (type,project_id,the_geom)
```

Comment:
quick note on this and in general as you progress through imports of geo data which in some real cases may span tens of thousands of records, time a few columns: let's consider both PostgreSQL's limits on query parameters (IIRC depends on versions, but easy to hit at the 10k scale, I think) and the efficiency of TypeORM at composing queries with "high" (for some value of high" numbers of params.

My general suggestion would be to open a transaction (if not in one already) and split the inserts into batches considering the number of parameters needed to safely set the batch size according to PostgreSQL's limits.

I also suggest to always add a short note about this in a comment next to the batching logic, so that whoever may update the query next and for example add more params could take that into account and see if it may be necessary to reduce the batch length according (e.g. if doubling the number of params needed for each row, halve the batch size)

Reply:
Wrapping the insert operations inside a transaction makes totally sense 👍 . However, as we are already working with streams the inserts are being made in batches (chunks) right?